File: Leetcode/All/199_Binary_Tree_Right_Side_View.java
public class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ret = new LinkedList<Integer>();
        if(root == null)    return ret;
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i = 0; i < size; i++){
                TreeNode temp = queue.poll();
                if(i == size-1)
                    ret.add(temp.val);
                if(temp.left != null)
                    queue.add(temp.left);
                if(temp.right != null)
                    queue.add(temp.right);
            }
        }
        return ret;
    }
}

File: Leetcode/All/199_Binary_Tree_Right_Side_View.cpp
class Solution {
public:
    int level = 0;
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ret;
        if(root == NULL)    return ret;
        inorder(root, 1, ret);
        return ret;
    }
    
    void inorder(TreeNode* root, int curLevel, vector<int>& ret){
        if(root == NULL)    return;
        if(curLevel > level){
            level = curLevel;
            ret.push_back(root->val);
        }
        inorder(root->right, curLevel + 1, ret);
        inorder(root->left, curLevel + 1, ret);
    }
};


class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ret;
        queue<TreeNode*> q;
        if(root != NULL)    
            q.push(root);
        while(!q.empty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                TreeNode* cur = q.front();
                if(i == size - 1)   ret.push_back(cur->val);
                if(cur->left != NULL)
                    q.push(cur->left);
                if(cur->right != NULL)
                    q.push(cur->right);
                q.pop();
            }
        }
        return ret;
    }
};